Userland/Libraries/LibGfx/VectorN.h

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2020, Stephan Unverwerth <s.unverwerth@serenityos.org>
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Array.h>
+#include <AK/Error.h>
+#include <AK/Format.h>
+#include <AK/Math.h>
+#include <AK/StdLibExtras.h>
+#include <AK/String.h>
+#include <AK/StringView.h>
+
+namespace Gfx {
+template<size_t N, typename T>
+requires(N >= 2u && N <= 4u) class VectorN final {
+public:
+    constexpr VectorN() = default;
+    constexpr VectorN(T x, T y) requires(N == 2)
+        : m_data { x, y }
+    {
+    }
+    constexpr VectorN(T x, T y, T z) requires(N == 3)
+        : m_data { x, y, z }
+    {
+    }
+    constexpr VectorN(T x, T y, T z, T w) requires(N == 4)
+        : m_data { x, y, z, w }
+    {
+    }
+
+    constexpr T x() const { return m_data[0]; }
+    constexpr T y() const { return m_data[1]; }
+    constexpr T z() const requires(N >= 3) { return m_data[2]; }
+    constexpr T w() const requires(N >= 4) { return m_data[3]; }
+
+    constexpr void set_x(T value) { m_data[0] = value; }
+    constexpr void set_y(T value) { m_data[1] = value; }
+    constexpr void set_z(T value) requires(N >= 3) { m_data[2] = value; }
+    constexpr void set_w(T value) requires(N >= 4) { m_data[3] = value; }
+
+    constexpr VectorN& operator+=(const VectorN& other)
+    {
+#ifdef __clang__
+#    pragma unroll 4
+#else
+#    pragma GCC unroll 4
+#endif
+        for (auto i = 0u; i < N; ++i) {
+            m_data[i] += other.m_data[i];
+        }
+        return *this;
+    }
+
+    constexpr VectorN& operator-=(const VectorN& other)
+    {
+#ifdef __clang__
+#    pragma unroll 4
+#else
+#    pragma GCC unroll 4
+#endif
+        for (auto i = 0u; i < N; ++i) {
+            m_data[i] -= other.m_data[i];
+        }
```

Comment:
ngl, this is more code than manually unrolling it
to compare:
```suggestion
        m_data[0] -= other.m_data[0];
        m_data[1] -= other.m_data[1];
        if constexpr (N > 2)
            m_data[2] -= other.m_data[2];
        if constexpr (N > 3)
            m_data[3] -= other.m_data[3];
```

Reply:
We lose flexibility in N > 4, but we'll probably never use that. Adding a static_assert on N <= 4 should be enough

Reply:
Sure, but I did not mean that we should use all of those ifdefs before each loop, see my macro above :sweat_smile:

Reply:
I get it now. That was similarly my issue. I couldn't use N as it was a template parameter. This makes the code a little "spaghetti-ish" due to the `4` being defined in multiple places. That can be somewhat limited though... It does make the code _look_ nicer.